### PR TITLE
Fix handoff isn't processed correctly

### PR DESCRIPTION
**Pull request by yah01**
Signed-off-by: yah01 <yang.cen@zilliz.com>
Co-authored-by: Congqi Xia <congqi.xia@zilliz.com>

#17627 #17416

### CODE REVIEW

File: internal/querynode/query_node.go [hunk 1]
```diff
@@ -448,3 +448,20 @@ func filterDuplicateChangeInfo(line *querypb.SegmentChangeInfo) {
 		line.OfflineSegments = filteredOfflineSegments
 	}
 }
+
+// isDuplicatedSegmentInfo checks whether online and offline segment are duplicated
+// checks segment id and segment states
+// to enable load balance between differenct query node, segment segment id with difference node ids are allowed
+func isDuplicatedSegmentInfo(online, offline *querypb.SegmentInfo) bool {
```
Comment (xiaofan-luan) [inline internal/querynode/query_node.go:455]: we should remove the duplicated segment info implementation

File: internal/querynode/shard_cluster.go [hunk 2]
```diff
@@ -394,18 +394,28 @@ func (sc *ShardCluster) init() {
 
 // pickNode selects node id in cluster
 func (sc *ShardCluster) pickNode(evt segmentEvent) (shardSegmentInfo, bool) {
-	for _, nodeID := range evt.nodeIDs {
+	nodeID, has := sc.selectNodeInReplica(evt.nodeIDs)
```
Comment (xiaofan-luan) [inline internal/querynode/shard_cluster.go:397]: need more log to explain what is happening

File: internal/querycoord/meta.go [hunk 3]
```diff
@@ -602,38 +602,38 @@ func (m *MetaReplica) saveGlobalSealedSegInfos(saves col2SegmentInfos) (col2Seal
...
-
-					for _, node := range offlineNodes {
-						segmentsChangeInfo.Infos = append(segmentsChangeInfo.Infos,
-							&querypb.SegmentChangeInfo{
-								OfflineNodeID:   node,
-								OfflineSegments: []*querypb.SegmentInfo{offlineInfo},
-							})
-					}
-				}
+				onlineInfo.NodeIds = diffSlice(info.NodeIds, offlineInfo.NodeIds...)
```
[the remark this answers is not in the document]
Comment (yah01) [reply internal/querycoord/meta.go:612]: info here is the new SegmentInfo whose NodeIds contains the new loaded node, and the offlineInfo is the old SegmentInfo. This diff make the onlineInfo's NodeIds contain only the new loaded node

File: internal/querycoord/meta.go [hunk 4]
```diff
@@ -602,38 +602,38 @@ func (m *MetaReplica) saveGlobalSealedSegInfos(saves col2SegmentInfos) (col2Seal
...
-					for _, node := range compactionSegmentInfo.NodeIds {
-						segmentsChangeInfo.Infos = append(segmentsChangeInfo.Infos, &querypb.SegmentChangeInfo{
-							OfflineNodeID:   node,
-							OfflineSegments: []*querypb.SegmentInfo{compactionSegmentInfo},
-						})
-					}
-					segmentsCompactionFrom = append(segmentsCompactionFrom, compactionSegmentInfo)
+				offlineInfo, err := m.getSegmentInfoByID(compactionSegmentID)
+				if err == nil && offlineInfo.SegmentState == commonpb.SegmentState_Sealed {
+					compactChangeInfo.OfflineSegments = append(compactChangeInfo.OfflineSegments, offlineInfo)
```
Comment (xige-16) [inline internal/querycoord/meta.go:628]: may be need (offlineInfo.NodeIds = diffSlice(offlineInfo.NodeIds, info.NodeIds...))

Comment (yah01) [reply internal/querycoord/meta.go:628]: The compacted segments aren't the same as above, we should offload them on all nodes

**Comment by yah01** [comment]
rerun ut

**Pull request merged**